### ISSUE TITLE
bug(`forge bind --alloy`): throws `unlinked bytecode with placeholder` error (forge build works)

### ISSUE THREAD

**Issue by supernovahs** [foundry-rs/foundry#9482]
### Component

Forge

### Have you ensured that all of these are up to date?

- [X] Foundry
- [ ] Foundryup

### What version of Foundry are you on?

forge 0.2.0 (31dd1f7 2024-11-27T00:22:49.993505000Z)

### What command(s) is the bug in?

forge bind --alloy 

### Operating System

None

### Describe the bug

unlinked bytecode with placeholder error (forge build works)

solc version 0.8.27

**Comment by zerosnacks** [comment]
Hi @supernovahs, thanks for flagging

Would you mind sharing more details about your setup or share the repo?

In the meantime, would you mind running `forge clean` and then `forge bind --alloy`?

**Comment by supernovahs** [comment]
Reproduce

Clone - https://github.com/Layr-Labs/eigenlayer-middleware
git checkout feat/slashing-release-branch

forge build 
forge bind --alloy

```
eigenlayer-middleware/out/BLSApkRegistry.sol/BLSApkRegistry.json:BLSApkRegistry

Context:
- Error #0: failed to parse input
- Error #1: invalid JSON: expected bytecode, found unlinked bytecode with placeholder: b77956a640ff14d8a7f287227aabbc1987 at line 1 column 37238

```

**Comment by zerosnacks** [comment]
This appears to be related to `BN254.sol` in `BLSApkRegistry` containing external function methods making the library to be considered an external library which requires more configuration. 
Alternatively, the easiest way, is to make the library only have `internal` methods unless this is explicitly intentioned.

The same is the case for `QuorumBitmapHistoryLib` and `SignatureCheckerLib`.

Configurable by in `foundry.toml`:

```
libraries = [
    "src/libraries/BN254.sol:BN254:0x0000000000000000000000000000000000000111",
"src/libraries/QuorumBitmapHistoryLib.sol:QuorumBitmapHistoryLib:0x0000000000000000000000000000000000000222",
    "src/libraries/SignatureCheckerLib.sol:SignatureCheckerLib:0x0000000000000000000000000000000000000333"  
]
```

See: https://book.getfoundry.sh/reference/forge/forge-build#linker-options

For reference:

```
2015181600101559050505f73__$b77956a640ff14d8a7f287227aabbc1987$__63574ae6ab866040518263fffff
```

**Comment by zerosnacks** [comment]
Marking as `resolved`, configuration issue

**Comment by sveitser** [comment]
Is there a replacement for this previous workflow with ethers bindings: 

1. Create bindings with unlinked libraries.
2. Link at runtime. 

Example: 

https://github.com/EspressoSystems/espresso-sequencer/blob/8b2babe3aacd459d626479df0238360d1570f7a8/utils/src/deployer.rs#L213

Currently it sounds like alloy bindings can only be generated if we link the library when generating the binaries but we would like to link different addresses on different chains.

**Comment by zerosnacks** [comment]
^ cc @yash-atreya / @DaniPopes

**Comment by DaniPopes** [comment]
I guess we could ignore the fact that the bytecode is unlinked and emit a warning when loading from JSON

**Comment by sveitser** [comment]
That would work for our use case. I'm curious though if I'm missing something and we're doing something non-idiomatic and happy to consider changing the workflow as we migrate to alloy.

Isn't it somewhat usual to not yet know the deployed contract addresses at the time bindings are generated, since the addresses will depend on accounts and nonces used to deploy the library contract?

**Comment by supernovahs** [comment]
Do we have an option now to ignore the fact that bytecode is unlinked?

**Comment by mattsse** [comment]
thanks @supernovahs 

investigating

**Comment by mattsse** [comment]
ah I see,

this currently enforces valid bytecode during parsing:

https://github.com/foundry-rs/foundry/blob/b53e6a9d9c3332303b83920f1d181f04752987d7/crates/sol-macro-gen/src/sol_macro_gen.rs#L34-L42

but unlinked has bytecode

I think we can work around this until properly support in alloy-sol

**Comment by sveitser** [comment]
Our team would be glad if generating bindings with unlinked contracts could be allowed. Currently we link placeholder addresses and replace them before deployment which seems a bit hacky and potentially error prone.

**Comment by mattsse** [comment]
do you need bindings for those contracts or can they be ignored?

**Comment by sveitser** [comment]
Yes we need them. I think our use case isn't so weird. We have a few contracts that split out some external libs to reduce contract size. We generate bindings and commit them to the repo. This has the advantage that team members who don't develop contracts don't need to install / run any EVM tools. So we just have one version of the bindings for all networks and need to link different addresses during deployment on different chains later.

**Comment by yash-atreya** [comment]
Reopening this as https://github.com/foundry-rs/foundry/pull/10291 is just a workaround and we should have a proper solution when https://github.com/alloy-rs/core/issues/926 is completed
